python/ray/serve/controller.py

```diff
@@ -129,8 +132,60 @@ def get_http_proxies(self) -> Dict[NodeId, ActorHandle]:
         """Returns a dictionary of node ID to http_proxy actor handles."""
         return self.http_state.get_http_proxy_handles()
 
+    async def update_autoscaling(self) -> None:
+        for deployment_name, (backend_info,
+                              route_prefix) in self.list_deployments().items():
+            backend_config = backend_info.backend_config
+            autoscaling_config = backend_config.autoscaling_config
+            if autoscaling_config is None:
+                continue
+            replicas = self.backend_state_manager._backend_states[
+                deployment_name]._replicas
+            running_replicas = replicas.get([ReplicaState.RUNNING])
+            current_num_ongoing_requests = []
+            for replica in running_replicas:
+                replica_tag = replica.replica_tag
+                num_ongoing_requests = (
+                    self.autoscaling_metrics_store.window_average(
+                        replica_tag,
+                        time.time() - autoscaling_config.look_back_period_s))
+                current_num_ongoing_requests.append(num_ongoing_requests)
+            if len(current_num_ongoing_requests) == 0:
+                print("length of list of num requests is zero")
+                continue
+            print("current ongoing num requests: ",
+                  current_num_ongoing_requests)
+            new_backend_config = backend_config.copy()
+            new_backend_config.num_replicas = calculate_desired_num_replicas(
+                autoscaling_config, current_num_ongoing_requests)
+            print("desired num replicas: ", new_backend_config.num_replicas)
+            print("Metrics dump:  ",
+                  self._dump_autoscaling_metrics_for_testing())
+            # continue #TODO: remove
+            replica_config = backend_info.replica_config
+            deployer_job_id = backend_info.deployer_job_id
+            backend_config_proto_bytes = new_backend_config.to_proto_bytes()
+            # TODO: Only allow autoscaling for versioned deployments for now
```

Comment:
This is because unversioned deployments tear down all the old replicas upon redeployment, even if you're just changing `num_replicas` by a small delta.  We will be able to support autoscaling for unversioned deployments with a small refactor in a future PR.